# 2-4 Intro Python Practice
## Working with Files

<font size="5" color="#00A0B2"  face="verdana"> <B>Student will be able to</B></font>  
 
- Import files into Jupyter notebooks  
- Open and **`.read()`** local files in memory   
- Use **`.read()`** to read a specific number of characters  
- Use **`.readlines()`** to read data from file as a **list** of lines  
- Use **`.readlines()`** to read data from file as a **list** of lines   
- Use **`.readline()`** to read data from file a line at a time   
- Use **`.strip()`** to remove new line characters  
- Use `.write()` to write data to a new local file  
- Use **`.seek()`** to set file read or write location  
- Use file append mode  


# &nbsp;
<font size="6" color="#B24C00"  face="verdana"> <B>Task 1</B></font>

## Order the Rainbow
#### Open the rainbow file then put in a list and print in alphabetical order
Download and open the file. 
- Download list of rainbow colors, as `rainbow.txt`, using `curl` 
  https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/rainbow   
- Open rainbow.txt in read mode using a variable: rainbow_file   
 

In [12]:
# [ ] import https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/rainbow as rainbow.txt


!curl https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/rainbow -o rainbow.txt

with open('rainbow.txt', 'r') as rainbow_file:
    rainbow_list = rainbow_file.read().splitlines()

rainbow_list.sort()

for color in rainbow_list:
    print(color)


blue
green
indigo
orange
red
violet
yellow


  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (35) schannel: next InitializeSecurityContext failed: Unknown error (0x80092012) - The revocation function was unable to check revocation for the certificate.



- Read rainbow_file as a list variable: rainbow_colors using `.readlines()`    


In [2]:
# [ ]  Open rainbow.txt in read mode & read as list with .readlines()


with open('rainbow.txt', 'r') as rainbow_file:
    rainbow_list = rainbow_file.readlines()

print(rainbow_list)

['red\n', 'orange\n', 'yellow\n', 'green\n', 'blue\n', 'indigo\n', 'violet\n']


 
1. Sort the rainbow_lines list alphabetically.  
2. Print each line of rainbow_lines by iterating the sorted list.  
3. Close rainbow_file.  

In [3]:
# [ ] sort rainbow_colors list, iterate the list to print each color

rainbow_list.sort()

for color in rainbow_list:
    print(color)
    

blue

green

indigo

orange

red

violet

yellow



# &nbsp;
<font size="6" color="#B24C00"  face="verdana"> <B>Task 2</B></font>

## The Weather
Create a program that reads from a file to display city name and average temperature in Celsius.   

&nbsp;  
- use `!curl` to download https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/world_temp_mean.csv as `mean_temp.txt`  
  


In [18]:
# [ ] The Weather: import world_mean_team.csv as mean_temp.txt

!curl https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/world_mean_temp -o mean_temp.txt


with open('mean_temp.txt', 'r') as mean_temp_file:
    mean_temp_list = mean_temp_file.readlines()

for line in mean_temp_list:
    print(line)

city,country,month ave: highest high,month ave: lowest low

Beijing,China,30.9,-8.4

Cairo,Egypt,34.7,1.2

London,UK,23.5,2.1

Nairobi,Kenya,26.3,10.5

New York City,USA,28.9,-2.8

Sydney,Australia,26.5,8.7

Tokyo,Japan,30.8,0.9



  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (35) schannel: next InitializeSecurityContext failed: Unknown error (0x80092012) - The revocation function was unable to check revocation for the certificate.


1. Open the file in `'r'` mode.  
2. Read the first line of text into a variable called: `  headings` and `print()`.  
3. Convert `headings` to a list using **`.split(',')`** which splits on each comma, `print()` the list.


In [27]:
# [ ] The Weather: open file, read/print first line, convert line to list (splitting on comma)

with open('mean_temp.txt', 'r') as mean_temp_file:
    mean_temp_list = mean_temp_file.readlines()

headings = mean_temp_list[0].split(',')
print(headings)

['city', 'country', 'month ave: highest high', 'month ave: lowest low\n']


#### use a while loop to read the remaining lines from the file  
  1. Assign remaining lines to a **`city_temp`** variable.  
  2. Convert the city_temp to a list using **`.split(',')`** for each **`.readline()`** in the loop.  
  3. Print each city & the highest monthly average temperature.  
  4. Close mean_temps.  

>Tips & Hints:   
- Use the print output of **`headings`** to determine the city_temp indexes to use.  
- "month ave: highest high" for Beijing is 30.9 Celsius.  
- Convert `city_temp` to lists with `.split(',')`.  

In [40]:
# [ ] The Weather: use while loop to print city and highest monthly average temp in celsius

with open('mean_temp.txt', 'r') as mean_temp_file:
    mean_temp_list = mean_temp_file.readlines()

city_temp = mean_temp_list[0].split(',')

city_index = city_temp.index('city')
celsius_index = city_temp.index('month ave: highest high')


mean_temp_list.pop(0)
while mean_temp_list:
    line = mean_temp_list.pop(0)
    line_list = line.split(',')
    print(city_temp[celsius_index] + " for " + line_list[city_index] + " is "+ line_list[celsius_index])


month ave: highest high for Beijing is 30.9
month ave: highest high for Cairo is 34.7
month ave: highest high for London is 23.5
month ave: highest high for Nairobi is 26.3
month ave: highest high for New York City is 28.9
month ave: highest high for Sydney is 26.5
month ave: highest high for Tokyo is 30.8


# &nbsp;  
<font size="6" color="#B24C00"  face="verdana"> <B>Task 3</B></font>  

## Random pi  guessing
Create random appearing numbers by reading digits of pi 
*Note: only "appears" random*

- Download https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/digits_of_pi  as pi.txt

In [41]:
# [ ] use curl to download https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/digits_of_pi as pi.txt

!curl https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/digits_of_pi -o pi.txt


  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (35) schannel: next InitializeSecurityContext failed: Unknown error (0x80092012) - The revocation function was unable to check revocation for the certificate.


#### Set up the project files and intitial values  
1. Open pi.txt in read mode, the file has a single line of text "3.14....".  
2. Get user name as input and say "hi".  
3. Use the length of `name` for variable called `seed`.  
4. Use `.seek()` with the value of `seed` to set the initial pointer location reading the file.  
5. Create a variable `digit` and assign it the value of reading one character from the file.  
6. Get `guess` variable value from users `input`  - "enter a single digit guess or "q" to quit".  
7. Initialize `correct` and `wrong` counter variables to **`0`**  (zero).  

In [52]:
# # [ ] Set up the project files and initial values

pi_file = open('pi.txt', 'r')

pi_list = pi_file.readlines()

name = input("Enter your name: ")
print("Hi " + name)

seed = len(name)
pi_file.seek(seed)

digit = pi_file.read(1)

guess = input("Enter a single digit guess or 'q' to quit: ")

correct = 0
wrong = 0

Hi 3


#### Create a while loop that tests that `guess` is a *digit* string
then in the loop:
1. if `digit` ( read from pi file) is "**.**" read the next character for digit  
2. else if `digit` is "\n" increment `seed` and use `seed` to set the pointer uing .`seek()`  
3. else see if `guess` is equal to `digit`   
  a. if `guess` equals `digit`: print "correct" and increment the varible named `correct`  
  b. if `guess` not equal `digit`: print "incorrect" and increment the variable named `wrong`  
  
**end the while loop** when user enters any non-digit(s) for `guess`, like "q".  
- Print `correct` and `wrong` values within a message to the user.  
- Close the pi file.  

In [53]:
while guess != 'q':
    if guess == digit:
        print("Correct")
        correct += 1
    else:
        print("Wrong")
        wrong += 1
    guess = input("Enter a single digit guess or 'q' to quit: ")
    digit = pi_file.read(1)

print("Correct: " + str(correct))
print("Wrong: " + str(wrong))

pi_file.close()


Wrong
Wrong
Wrong
Correct
Wrong
Correct
Wrong
Correct
Correct
Wrong
Correct: 4
Wrong: 6


[Terms of use](http://go.microsoft.com/fwlink/?LinkID=206977) &nbsp; [Privacy & cookies](https://go.microsoft.com/fwlink/?LinkId=521839) &nbsp; © 2017 Microsoft